src/ResourceHelperTrait.php

```diff
@@ -0,0 +1,95 @@
+<?php
+/*
+ * Copyright 2022 Google LLC
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace Google\ApiCore;
+
+use Google\ApiCore\ValidationException;
+
+trait ResourceHelperTrait
+{
+    private static $templateMap;
+
+    // Must be implemented by extendees to call loadTemplates.
+    private static function registerTemplates()
+    {
+        // TODO: Add void return type hint.
+        self::$templateMap = [];
+    }
+
+    private static function loadTemplates(string $configPath, string $serviceName)
+    {
+        // TODO: Add void return type hint.
+        if (!is_null(self::$templateMap)) {
+            return;
+        }
+
+        $descriptors = require($configPath);
+        $templates = $descriptors['interfaces'][$serviceName]['templateMap'] ?? [];
+        self::$templateMap = [];
+        foreach ($templates as $name => $template) {
+            self::$templateMap[$name] = new PathTemplate($template);
+        }
+    }
+
+    private static function getTemplate(string $key)
```

Comment:
The function name `getTemplate` exists on one gapic client:

 - [`Dataflow/src/V1beta3/Gapic/TemplatesServiceGapicClient.php`](https://github.com/googleapis/google-cloud-php/blob/main/Dataflow/src/V1beta3/Gapic/TemplatesServiceGapicClient.php#L321)

Note: this will not be a conflict once we use the `send` prefix, but we should think of a more explicit name in the meantime, e.g. `getPathTemplate` (I've confirmed this does not exist on any clients)

Note: This also means we would want to rename `loadTemplates` and `registerTemplates` to be consistent with the naming that we choose for `getTemplate`

Reply:
Great call, I'll do that. Done.